drivers/spi/spi_xec_gpspi.c

```diff
@@ -0,0 +1,791 @@
+
+	/* soft reset is self clearing */
+	regs->control |= BIT(XEC_GPSPI_CTRL_SRST_POS);
+
+	regs->clock_gen = clk_gen;
+	regs->clock_control = clk_ctrl;
+}
+
+/* Calculate GPSPI frequency divider register field value based upon
+ * the configured GPSPI reference clock frequency: 48 or 2 MHz.
+ * SPI clock frequency = (reference_clock) / (2 * preload)
+ * preload 0 is a special value. When preload is 0 the hardware forces
+ * SPI clock to be 48 MHz.
+ */
+static int gpspi_configure_spi_clock(struct xec_gpspi_regs * const regs, uint32_t spi_clk_hz)
+{
+	uint32_t preload, ref_clk;
+
+	if (spi_clk_hz > MHZ(24)) {
+		/* HW can only do 48 MHz. set preload = 0 */
+		preload = 0u;
+	} else if (spi_clk_hz < KHZ(16)) {
+		regs->clock_control &= ~BIT(XEC_GPSPI_CLK_CTRL_SRC_CLK_48M_POS);
+		preload = 63u;
+	} else {
+		regs->clock_control &= ~BIT(XEC_GPSPI_CLK_CTRL_SRC_CLK_48M_POS);
+		ref_clk = MHZ(2);
+		if (spi_clk_hz > ref_clk) {
+			ref_clk = MHZ(48);
+			regs->clock_control |= BIT(XEC_GPSPI_CLK_CTRL_SRC_CLK_48M_POS);
+		}
+
+		preload = ref_clk / (2u * spi_clk_hz);
+	}
+
+	regs->clock_gen = preload;
+
+	return 0;
+}
+
+
+static const uint8_t gpspi_spi_mode_tbl[] = {
+	XEC_GPIO_CLK_CTRL_SPI_MODE_0, XEC_GPIO_CLK_CTRL_SPI_MODE_1,
+	XEC_GPIO_CLK_CTRL_SPI_MODE_2, XEC_GPIO_CLK_CTRL_SPI_MODE_3
+};
+
+static int gpspi_set_spi_mode(const struct device *dev, const struct spi_config *spi_conf)
+{
+	const struct spi_xec_gpspi_config * const cfg = dev->config;
+	struct xec_gpspi_regs * const regs = cfg->regs;
+	uint8_t index = 0;
+
+	if (spi_conf->operation & SPI_MODE_CPHA) {
+		index = 1u;
+	}
+	if (spi_conf->operation & SPI_MODE_CPOL) {
+		index += 2u;
+	}
+
+	regs->clock_control &= ~(XEC_GPSPI_CLK_CTRL_PH_MSK);
+	regs->clock_control |= gpspi_spi_mode_tbl[index & 0x3u];
+
+	return 0;
+}
+
+/* NOTE: struct spi_config.operation is defined as uint32_t when
+ * CONFIG_SPI_EXTENDED_MODES is enabled otherwise it is uint16_t.
+ * SPI_LINES_xxx are located at bit 16.
+ */
+static int gpspi_check_unsupported_features(const struct spi_config *spi_conf)
+{
+	uint32_t lines = spi_conf->operation & SPI_LINES_MASK;
+
+	if (spi_conf->operation & (SPI_OP_MODE_SLAVE | SPI_MODE_LOOP)) {
+		LOG_ERR("Does not support SPI device or loop back");
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_SPI_EXTENDED_MODES)) {
+		if (lines != SPI_LINES_SINGLE) {
+			LOG_ERR("Supports single(full-duplex) mode only");
+			return -ENOTSUP;
+		}
+	}
+
+	if (spi_conf->operation & SPI_CS_ACTIVE_HIGH) {
+		LOG_ERR("CS active high not supported");
+		return -ENOTSUP;
+	}
+
+	if (SPI_WORD_SIZE_GET(spi_conf->operation) != 8) {
+		LOG_ERR("Word size != 8 not supported");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int req_full_reconfig(const struct device *dev, const struct spi_config *spi_conf)
+{
+	struct spi_xec_gpspi_data * const data = dev->data;
+
+	if (data->scfg.frequency != spi_conf->frequency) {
+		return 1; /* frequency change */
+	} else if ((data->scfg.operation & SPI_MODE_MASK)
```

Comment:
why not coalescing the 2 tests in one if since you return the same value on both?